### DISCUSSION TITLE
Unable to get clean Schema on filter hooks context

### DISCUSSION THREAD

**Discussion by andres-torres-epam** [directus/directus#15331] [Archive]
### Describe the Bug

The Extension Context will hang (not return) if you try to await to get the schema after a call to the cache clear inside a 'collections.delete' filter hook.

P.S. I think this is related to the fact that the filter hook is called within a knex transaction inside the ItemsService. The cache clear could be making that transaction reach an undefined behavior.

### To Reproduce

Create a filter hook that clears the cache using an http post request to the endpoint /utils/cache/clear and then try to await for a call to the context.getSchema() method to complete; you will see the process hangs without being able to complete.

### What version of Directus are you using?

9.16.1

### What version of Node.js are you using?

16.15.1

### What database are you using?

Postgres

### What browser are you using?

Chrome

### How are you deploying Directus?

running locally

**Comment by rijkvanzanten** [comment]
Can you share a reproduction of the hook? I'm pretty sure it's what you mentioned: `getSchema()` reads the schema outside of the current transaction, which makes the database transaction hang, but we'll have to look at the used code to confirm 👍🏻 

Filters run in the transaction context, actions run outside of it

**Reply by andres-torres-epam** [reply]
```js
function cleanCache() {
  return axios.post(`http://${process.env.HOST}:${process.env.PORT}/utils/clean/cache?access_token=${my_secure_token}`);
}

function newProvider(context) {
  return async function(service) {
    await cleanCache();
    const schema = await context.getSchema(); // This will never complete execution
    return new context.services[service]({database: context.database, schema });
  }
}

export default defineHook((register, context) => {
  const provider = newProvider(context);
  register.filter('collections.delete', async (payload) => {
    const collectionsService = await provider('CollectionsService');
    // Do things with service...
  })
});
```

I've been thinking about a solution, and what a I came up with could be to use a mutex (blocking state) that makes the request to the clean cache fail with a descriptive error and add a note to the documentation explaining this behavior when executing within a filter hook context.

**Reply by rijkvanzanten** [reply]
Could you try passing the current database instance to the getSchema function? That way it should re-use the transaction and not lock-up 🤔 

`await context.getSchema({ database: context.database })`

**Reply by andres-torres-epam** [reply]
Even if you pass in the database it fails to complete...

**Reply by rijkvanzanten** [reply]
What are you trying to do with the service exactly? You might be better of by using an `action` instead of a `filter` so you're not bound to the current transaction context

**Reply by andres-torres-epam** [reply]
I fixed it for my current requirement (used a filter hook with a provider that doesn't clear the cache). This thread is meant to work as an informative / possible issue report. Sorry if it wasn't clear.
